rasa/engine/storage/storage.py

```diff
@@ -0,0 +1,149 @@
+            resource: The resource which should be persisted.
+
+        Returns:
+            A directory which can be used to persist data for the given `Resource`.
+        """
+        ...
+
+    @contextmanager
+    @abc.abstractmethod
+    def read_from(self, resource: Resource) -> ContextManager[Path]:
+        """Provides the data of a persisted `Resource`.
+
+        Args:
+            resource: The `Resource` whose persisted should be accessed.
+
+        Returns:
+            A directory containing the data of the persisted `Resource`.
+
+        Raises:
+            ValueError: In case no persisted data for the given `Resource` exists.
+        """
+        ...
+
+    def create_model_package(
+        self,
+        model_archive_path: Union[Text, Path],
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        domain: Domain,
+    ) -> None:
+        """Creates a model archive containing all data to load and run the model.
+
+        Args:
+            model_archive_path: The path to the archive which should be created.
+            train_schema: The schema which was used to train the graph model.
+            predict_schema: The schema for running predictions with the trained model.
+            domain: The `Domain` which was used to train the model.
+        """
+        ...
+
+
+@dataclass()
+class ModelMetadata:
+    """Describes a trained model."""
+
+    trained_at: datetime
+    rasa_open_source_version: Text
+    model_id: Text
+    domain: Domain
+    train_schema: GraphSchema
+    predict_schema: GraphSchema
+
+    def as_dict(self) -> Dict[Text, Any]:
+        """Returns serializable version of the `ModelMetadata`."""
+        return {
+            METADATA_DOMAIN_KEY: self.domain.as_dict(),
+            METADATA_TRAINED_AT_KEY: self.trained_at.isoformat(),
+            METADATA_MODEL_ID: self.model_id,
+            METADATA_VERSION_KEY: self.rasa_open_source_version,
+            METADATA_TRAIN_SCHEMA: self.train_schema.as_dict(),
+            METADATA_PREDICT_SCHEMA: self.predict_schema.as_dict(),
+        }
+
+    @classmethod
+    def from_dict(cls, serialized: Dict[Text, Any]) -> ModelMetadata:
+        """Loads `ModelMetadata` which has been serialized using `metadata.as_dict()`.
+
+        Args:
+            serialized: Serialized `ModelMetadata` (e.g. read from disk).
+
+        Returns:
+            Instantiated `ModelMetadata`.
+        """
+        metadata = serialized.copy()
+        metadata[METADATA_TRAINED_AT_KEY] = datetime.fromisoformat(
+            metadata[METADATA_TRAINED_AT_KEY],
+        )
+        metadata[METADATA_DOMAIN_KEY] = Domain.from_dict(metadata[METADATA_DOMAIN_KEY])
+        metadata[METADATA_TRAIN_SCHEMA] = GraphSchema.from_dict(
+            metadata[METADATA_TRAIN_SCHEMA]
+        )
+        metadata[METADATA_PREDICT_SCHEMA] = GraphSchema.from_dict(
+            metadata[METADATA_PREDICT_SCHEMA]
+        )
+
+        return ModelMetadata(**metadata)
```

Comment:
This only works because the key constants match the dataclass attributes. Could we either:
1. have a single source of truth i.e. get the key names from the attributes. Or
2. do this line using keyword args directly.
?

Reply:
It just seems like redundancy. We already have the fields defined in the dataclass, why do we also have to have them as globally accessible constants at the top of this file?

Reply:
done